cpp/benchmarks/bench_core.h

```diff
@@ -0,0 +1,266 @@
+            dt_max  = std::numeric_limits<double>::max();
+            f       = [model](Eigen::Ref<const Eigen::VectorXd> y_f, double t_f, Eigen::Ref<Eigen::VectorXd> dydt_f) { model.eval_right_hand_side(y_f, y_f, t_f, dydt_f); };
+            yt      = Eigen::VectorXd::Zero(model.get_initial_values().size());
+            ytp1    = Eigen::VectorXd::Zero(yt.size());
+            // set values for yt
+            std::vector<double> yt_vals({6377.873644, 35.249156, 30.029611, 182.145865, 66.153059, 79.530621, 3069.383604, 159.634440});
+            assert(yt.size() <= yt_vals.size());
+            for (int i = 0; i < yt.size(); i++) {
+                yt[i] = yt_vals[i];
+            }
+            return model;
+        }
+    };
+
+    template <char const *filepath, class Model=mio::benchmark::model::SecirAgeres>
+    struct integrator_step_file_init
+    {
+        typename Model::type operator () (
+            double& abs_tol, double& rel_tol, double& dt_min, double& dt_max,
+            double& t_init, double& dt_init,
+            mio::DerivFunction& f, Eigen::VectorXd& yt, Eigen::VectorXd& ytp1
+        ) {
+            // read configuration file
+            mio::benchmark::detail::ConfigReader config(filepath);
+            size_t num_agegroups;
+            config.read(num_agegroups);
+            config.read(t_init);
+            config.read(dt_init);
+            config.read(abs_tol);
+            config.read(rel_tol);
+            config.read(dt_min);
+            config.read(dt_max);
+            // create model and read initial yt values
+            typename Model::type model = Model()(num_agegroups);
+            std::vector<double> yt_vals(model.get_initial_values().size());
+            config.read(yt_vals);
+            // initialize remaining components
+            f     = [model](Eigen::Ref<const Eigen::VectorXd> y_f, double t_f, Eigen::Ref<Eigen::VectorXd> dydt_f) { model.eval_right_hand_side(y_f, y_f, t_f, dydt_f); };
+            yt    = Eigen::VectorXd::Zero(yt_vals.size());
+            ytp1  = Eigen::VectorXd::Zero(yt_vals.size());
+            for (int i = 0; i < yt.size(); i++) {
+                yt[i] = yt_vals[i];
+            }
+            return model;
+        }
+    };
+
+    template <class Integrator, class Initializer=default_integrator_step_init<>>
+    void integrator_step(::benchmark::State& state)
+    {
+        // suppress non-critical messages
+        mio::set_log_level(mio::LogLevel::critical);
+        // benchmark setup
+        double abs_tol, rel_tol, dt_min, dt_max, t_init, dt_init;
+        mio::DerivFunction f;
+        Eigen::VectorXd yt, ytp1;
+
+        auto model = Initializer()(abs_tol, rel_tol, dt_min, dt_max, t_init, dt_init, f, yt, ytp1);
+
+        auto I = Integrator(abs_tol, rel_tol, dt_min, dt_max);
+
+        double t, dt;
+        for (auto _ : state) {
+            // This code gets timed
+            mio::benchmark::detail::reset_time(t, dt, t_init, dt_init);

```

Comment:
why not just declare t and dt inside the loop body?

Reply:
Good question. reset_time used to do a lot more, and I just missed the point when it was no longer usefull as a function. Now removed